### PR TITLE
Add unit tests for DefaultAuthenticator role resolution logic

### LINKED ISSUE #5092: Add unit tests for DefaultAuthenticator role resolution logic
## Summary

`DefaultAuthenticator.extractRequestedRoles()` and `resolvePrincipalRoles()` contain non-trivial logic for parsing scope strings, filtering by the `PRINCIPAL_ROLE:` prefix, and resolving active principal roles, but currently lack dedicated unit test coverage.

The existing `DefaultAuthenticatorTest` verifies principal lookup and error handling, but does not exercise the role extraction and resolution code paths.

## Scope

Add unit tests covering:

- `extractRequestedRoles()`
  - `PRINCIPAL_ROLE:ALL` returns the "all roles" semantic
  - extraction of individual role names
  - filtering of scopes with invalid prefixes
  - mixed valid and invalid scopes
  - empty or missing role scopes

- `resolvePrincipalRoles()`
  - activating all granted roles when `ALL` is requested
  - activating only the requested roles when specific roles are provided
  - ignoring requested roles that are not granted to the principal
  - handling empty requested-role sets

## Why

These methods were recently modified and contain branching logic with several edge cases, including invalid scope prefixes and the special handling of `ALL` via empty-set semantics. Adding targeted unit tests will improve coverage and help prevent future regressions.

### To Reproduce

_No response_

### Actual Behavior

_No response_

### Expected Behavior

_No response_

### Additional context

_No response_

### System information

_No response_

### PR DESCRIPTION
**Pull request by psy-ns**
This PR adds dedicated, isolated unit test coverage for the protected extractRequestedRoles and resolvePrincipalRoles methods in DefaultAuthenticator, addressing Issue #5092.

While behavioral coverage existed via the main authenticate() method, these localized tests act as direct regression guards for the role resolution logic itself.

Key additions in DefaultAuthenticatorTest.java:

extractRequestedRoles tests: Verified prefix stripping (PRINCIPAL_ROLE:), handling of the :ALL flag, filtering of unprefixed invalid scopes, and empty credential scopes.

resolvePrincipalRoles tests: Verified integration with the metastore using the pre-created principalEntity, ensuring that valid subsets and the :ALL flag return correctly granted roles, while requested but ungranted roles strictly throw an AuthenticationFailedException.

## Checklist
- [x] 🛡️ Don't disclose security issues! (contact security@apache.org)
- [x] 🔗 Clearly explained why the changes are needed, or linked related issues: Fixes #5092 
- [x] 🧪 Added/updated tests with good coverage, or manually tested (and explained how)
- [x] 💡 Added comments for complex logic
- [ ] 🧾 Updated `CHANGELOG.md` (if needed)
- [ ] 📚 Updated documentation in `site/content/in-dev/unreleased` (if needed)

### CODE REVIEW

**Review by adutra — APPROVED** [review]

File: runtime/service/src/test/java/org/apache/polaris/service/auth/DefaultAuthenticatorTest.java [hunk 1]
```diff
@@ -518,4 +518,158 @@ private void assertUnauthorized(PolarisCredential credentials) {
...
+            authenticator.extractRequestedRoles(credentials);
+
+    assertThat(selection.allRolesRequested()).isFalse();
+    assertThat(selection.roles()).isEmpty();
+  }
+
+  // --- Tests for resolvePrincipalRoles ---
+
+  @Test
+  void testResolvePrincipalRoles_allRequestedReturnsAllGranted() {
```
Comment (flyingImer) [inline runtime/service/src/test/java/org/apache/polaris/service/auth/DefaultAuthenticatorTest.java:604]: These pin how allRolesRequested gets computed from the credential. Nothing pins how it survives resolvePrincipalAttributes. I'd suggest to add a new test case:
- Given: credentials naming one specific role, plus an incoming identity that already carries PRINCIPAL_ROLE_ALL_ATTRIBUTE_KEY set to true
- When: authenticate
- Then: the flag on the result is false and the roles are just that one role.

Worth covering PRINCIPAL_ENTITY_ATTRIBUTE_KEY the same way. Same overwrite path, and the resolved PrincipalEntity is what authorization reads downstream.

**Review by flyingImer — COMMENTED** [review]
These are the right two methods to pin down. Both only had coverage through authenticate() before.

One more case is worth adding. It is on the attribute map, not the role set. resolvePrincipalAttributes copies every incoming identity attribute in first. It writes Polaris's own reserved keys second. Then it calls buildKeepingLast, and that's the part worth pausing on. Guava's build() throws on a duplicate key. buildKeepingLast trades that hard failure for a silent overwrite, so if an incoming attribute happens to use the same string as PRINCIPAL_ROLE_ALL_ATTRIBUTE_KEY, Polaris's later write wins instead of the build blowing up. That's the right outcome.

Resolver reads that same key later. True means resolve every role the principal has been granted. False means resolve only the roles the credential named. One boolean, two different privilege outcomes.

So the whole guarantee here is the order of two statements. Nothing documents that the order matters. No test asserts it. Reorder those two lines in a future refactor and every test in this file still passes.

To be clear, this isn't a bug today. One writer of that key, one reader, correct order. I just want a test that fails the day someone changes it (future regression guardrail).

**Comment by adutra** [comment]
> One more case is worth adding. It is on the attribute map, not the role set. resolvePrincipalAttributes copies every incoming identity attribute in first. It writes Polaris's own reserved keys second. Then it calls buildKeepingLast, and that's the part worth pausing on. Guava's build() throws on a duplicate key. buildKeepingLast trades that hard failure for a silent overwrite, so if an incoming attribute happens to use the same string as PRINCIPAL_ROLE_ALL_ATTRIBUTE_KEY, Polaris's later write wins instead of the build blowing up. That's the right outcome.

Again, I've already offered to remove this bulk copy, which I introduced simply out of convenience. And I'm not opposed to switching to `build()` instead of `buildKeepingLast()`.

**Comment by adutra** [comment]
> > One more case is worth adding. It is on the attribute map, not the role set. resolvePrincipalAttributes copies every incoming identity attribute in first. It writes Polaris's own reserved keys second. Then it calls buildKeepingLast, and that's the part worth pausing on. Guava's build() throws on a duplicate key. buildKeepingLast trades that hard failure for a silent overwrite, so if an incoming attribute happens to use the same string as PRINCIPAL_ROLE_ALL_ATTRIBUTE_KEY, Polaris's later write wins instead of the build blowing up. That's the right outcome.
> 
> Again, I've already offered to remove this bulk copy, which I introduced simply out of convenience. And I'm not opposed to switching to `build()` instead of `buildKeepingLast()`.

https://github.com/apache/polaris/pull/5254

**Comment by flyingImer** [comment]
> > > One more case is worth adding. It is on the attribute map, not the role set. resolvePrincipalAttributes copies every incoming identity attribute in first. It writes Polaris's own reserved keys second. Then it calls buildKeepingLast, and that's the part worth pausing on. Guava's build() throws on a duplicate key. buildKeepingLast trades that hard failure for a silent overwrite, so if an incoming attribute happens to use the same string as PRINCIPAL_ROLE_ALL_ATTRIBUTE_KEY, Polaris's later write wins instead of the build blowing up. That's the right outcome.
> > 
> > 
> > Again, I've already offered to remove this bulk copy, which I introduced simply out of convenience. And I'm not opposed to switching to `build()` instead of `buildKeepingLast()`.
> 
> #5254

Thanks, #5254 removes the SecurityIdentity attribute bulk copy, so the overwrite/collision case I asked for is no longer part of this flow. My inline is obsolete and can be resolved.

File: runtime/service/src/test/java/org/apache/polaris/service/auth/DefaultAuthenticatorTest.java [hunk 2]
```diff
@@ -518,4 +518,158 @@ private void assertUnauthorized(PolarisCredential credentials) {
...
   }
-}
+  // --- Tests for extractRequestedRoles ---
+
+  @Test
+  void testExtractRequestedRoles_allReturnsEmptyWithFlag() {
+    PolarisCredential credentials =
+            PolarisCredential.of(null, PRINCIPAL_NAME, Set.of(DefaultAuthenticator.PRINCIPAL_ROLE_ALL));
+    DefaultAuthenticator.PrincipalRoleSelection selection =
+            authenticator.extractRequestedRoles(credentials);
```
Comment (flyingImer) [inline runtime/service/src/test/java/org/apache/polaris/service/auth/DefaultAuthenticatorTest.java:528]: I need to revise my earlier comment that these were the right two methods to pin down directly: I don't think this whole block should pin the helpers directly unless the current helper boundary and PrincipalRoleSelection shape are an intended subclass contract. 

The existing authenticate() tests already cover the caller-visible all/subset/empty/unprefixed/ungranted outcomes, so these direct calls mainly add coupling to that helper shape. 

If subclasses are expected to rely on it, could we make that contract explicit and cover it with a compact parameterized test? Otherwise, one authenticate() test for PRINCIPAL_ROLE:ALL taking precedence over a named role seems to cover the only missing caller-visible behavior.

**Review by flyingImer — COMMENTED** [review]
I don't think this is ready to merge as-is. Please either shrink the PR to behavior not already covered through authenticate(), or clarify and compact the protected-helper contract this is meant to stabilize; then get :polaris-runtime-service:check green against current main. With those outcomes, no blocking concerns from my side.
